test/e2e/kappcontroller/imgpkg_bundle_test.go

```diff
@@ -123,8 +118,226 @@ spec:
+		pkgiYaml := fmt.Sprintf(`---
+apiVersion: data.packaging.carvel.dev/v1alpha1
+kind: Package
+metadata:
+  name: pkg.test.carvel.dev.1.0.0
+  namespace: %[1]s
+  annotations:
+    kapp.k14s.io/change-group: "package"
+spec:
+  refName: pkg.test.carvel.dev
+  version: 1.0.0
+  template:
+    spec:
+      fetch:
+      - imgpkgBundle:
+          image: registry-svc.%[3]s.svc.cluster.local:5050/secret-test/test-bundle@sha256:387e2fa04e3c9d7b89f03be4fd3fb661fa3496352ece38a7fc8cba6780c7131d
+      template:
+      - ytt: {}
+      - kbld:
+          paths:
+          - "-"
+          - ".imgpkg/images.yml"
+      deploy:
+      - kapp: {}
+---
+apiVersion: packaging.carvel.dev/v1alpha1
+kind: PackageInstall
+metadata:
+  name: %[2]s
+  namespace: %[1]s
+  annotations:
+    kapp.k14s.io/change-group: kappctrl-e2e.k14s.io/packageinstalls
+    kapp.k14s.io/change-rule: "upsert after upserting package"
+spec:
+  syncPeriod: 30s
+  serviceAccountName: kappctrl-e2e-ns-sa
+  packageRef:
+    refName: pkg.test.carvel.dev
+    versionSelection:
+      constraints: 1.0.0
+`, env.Namespace, pkgiName, registryNamespace) + sas.ForNamespaceYAML()
+
+		kapp.RunWithOpts([]string{"deploy", "-a", pkgiName, "-f", "-"}, e2e.RunOpts{
+			StdinReader:  strings.NewReader(pkgiYaml),
+			OnErrKubectl: []string{"get", "app", pkgiName, "-oyaml"},
+		})
+
+		kubectl.Run([]string{"get", "configmap", "e2e-test-map"})
+		output := kubectl.Run([]string{"get", "app", pkgiName, "-oyaml"})
+		var appcr v1alpha1.App
+		require.NoError(t, yaml.Unmarshal([]byte(output), &appcr))
+		require.Equal(t, 0, appcr.Status.Fetch.ExitCode)
+		initialFetchUpdate = appcr.Status.Fetch.UpdatedAt
+	})
+	defer kapp.Run([]string{"delete", "-a", pkgiName})
+
+	logger.Section("remove registry", func() {
+		kapp.Run([]string{"delete", "-a", registryName, "-n", registryNamespace})
+		isRegistryPresent = false
+	})
+
+	logger.Section("wait for reconciliation and ensure no error occur while fetching", func() {
+		waitForPKGIReconciliationOrFail(t, kubectl, pkgiName, initialFetchUpdate)
+
+		var appcr v1alpha1.App
+		output := kubectl.Run([]string{"get", "app", pkgiName, "-oyaml"})
+		require.NoError(t, yaml.Unmarshal([]byte(output), &appcr))
+		require.Equal(t, 0, appcr.Status.Fetch.ExitCode)
+	})
+}
+
+func Test_PackageRepo_FetchFromCaching(t *testing.T) {
+	env := e2e.BuildEnv(t)
+	logger := e2e.Logger{}
+	kapp := e2e.Kapp{t, env.Namespace, logger}
+	kubectl := e2e.Kubectl{t, env.Namespace, logger}
+	sas := e2e.ServiceAccounts{env.Namespace}
+	isRegistryPresent := false
+
+	// If this changes, the skip-tls-verify domain must be updated to match
+	pkgrName := "pkgr-to-test-caching"
+	registryNamespace := "registry"
+	registryName := "test-registry"
+
+	cleanUp := func() {
+		// Delete App with kubectl first since kapp
+		// deletes ServiceAccount before App
+		kapp.Run([]string{"delete", "-a", pkgrName})
+		kapp.Run([]string{"delete", "-a", "secret-export"})
+		if isRegistryPresent {
```

Comment:
`if isRegistryPresent {` is not needed since kapp delete is idempotent. it will succeed if the app does not exist.